### ISSUE TITLE
AttributeError: 'Interpolator' object has no attribute 'frozen_assembled_interpolator'

### ISSUE THREAD

**Issue by ScottMacLachlan** [firedrakeproject/firedrake#1712]
When running in parallel, for certain (reproducible, but not consistent) problem sizes and parallel distributions, I trigger the error messages below.  Code that produces them is attached, and I can provide details on MPI distribution and git hashes as needed.

```Traceback (most recent call last):
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/firedrake/firedrake/interpolation.py", line 80, in interpol
ate
    assembled_interpolator = self.frozen_assembled_interpolator
AttributeError: 'Interpolator' object has no attribute 'frozen_assembled_interpolator'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/caching.py", line 197, in __new__
    return cls._cache_lookup(key)
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/caching.py", line 205, in _cache_lookup
    return cls._cache[key]
KeyError: ('46ca45b04e37cb1f437a6ae8e50f3de5', <class 'pyop2.base.Set'>, False, False, False, <class 'pyop2.base.Arg'>, <
Access.WRITE: 2>, (<class 'pyop2.petsc_base.Dat'>, dtype('float64'), (<class 'pyop2.petsc_base.DataSet'>, (1,), (<class '
pyop2.base.Set'>,))), ((<class 'pyop2.base.Map'>, 6, None),), False, 0, <class 'pyop2.base.Arg'>, <Access.READ: 1>, (<cla
ss 'pyop2.petsc_base.Dat'>, dtype('float64'), (<class 'pyop2.petsc_base.DataSet'>, (3,), (<class 'pyop2.base.Set'>,))), (
(<class 'pyop2.base.Map'>, 4, None),), False, 1, <class 'pyop2.base.Arg'>, <Access.READ: 1>, (<class 'pyop2.petsc_base.Gl
obal'>, dtype('float64'), (1,)), None, False, <class 'pyop2.base.Arg'>, <Access.READ: 1>, (<class 'pyop2.petsc_base.Globa
l'>, dtype('float64'), (1,)), None, False, <class 'pyop2.base.Arg'>, <Access.READ: 1>, (<class 'pyop2.petsc_base.Global'>
, dtype('float64'), (1,)), None, False, <class 'pyop2.base.Arg'>, <Access.READ: 1>, (<class 'pyop2.petsc_base.Global'>, d
type('float64'), (1,)), None, False, None, <class 'pyop2.sequential.JITModule'>, 4)
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/compilation.py", line 191, in compiler_version
    return Compiler.compiler_versions[self._cc]
KeyError: 'mpicc'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "3dgenerator.py", line 379, in <module>
    main()
  File "3dgenerator.py", line 311, in main
    bcs = problem.bcs(Z)
  File "3dgenerator.py", line 147, in bcs
    DirichletBC(Z.sub(1), as_vector([0, B_y_gen, 0]), [5,6]),
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/firedrake/firedrake/adjoint/dirichletbc.py", line 16, in wr
apper
    init(self, *args, **kwargs)
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/firedrake/firedrake/bcs.py", line 282, in __init__
    self.function_arg = g
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/firedrake/firedrake/bcs.py", line 349, in function_arg
    g = function.Function(self._function_space).interpolate(g)
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/firedrake/firedrake/function.py", line 353, in interpolate
    return interpolation.interpolate(expression, self, subset=subset)
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/firedrake/firedrake/interpolation.py", line 36, in interpol
ate
    return Interpolator(expr, V, subset=subset, access=access).interpolate()
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/firedrake/firedrake/interpolation.py", line 83, in interpol
ate
    assembled_interpolator = self.callable()
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/firedrake/firedrake/interpolation.py", line 174, in callabl
e
    l()
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/base.py", line 3561, in compute
    fun = self._jitmodule
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/utils.py", line 62, in __get__
    obj.__dict__[self.__name__] = result = self.fget(obj)
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/sequential.py", line 206, in _jitmodule
    pass_layer_arg=self._pass_layer_arg)
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/caching.py", line 199,--------------------------------------------------------------------------
MPI_ABORT was invoked on rank 192 in communicator MPI_COMM_WORLD
with errorcode 1.
NOTE: invoking MPI_ABORT causes Open MPI to kill all MPI processes.
You may or may not see output from other processes, depending on
exactly when Open MPI kills them.
--------------------------------------------------------------------------
 in __new__
    obj = make_obj()
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/caching.py", line 189, in make_obj
    obj.__init__(*args, **kwargs)
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/sequential.py", line 101, in __init__
    self.compile()
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/sequential.py", line 158, in compile
    comm=self.comm)
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/compilation.py", line 478, in load
    compiler = LinuxCompiler(cppargs, ldargs, cpp=cpp, comm=comm)
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/compilation.py", line 410, in __init__
    cpp=cpp, comm=comm)
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/compilation.py", line 185, in __init__
    self._cppargs = cppargs + configuration['cflags'].split() + self.workaround_cflags
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/compilation.py", line 203, in workaround_cflags
    compiler, ver = self.compiler_version
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/compilation.py", line 194, in compiler_version
    ver = sniff_compiler_version(self._cc)
  File "/gpfs/fs1/home/s/scottmac/scottmac/test/firedrake/src/PyOP2/pyop2/compilation.py", line 69, in sniff_compiler_version
    ver = subprocess.check_output([cc, "--version"]).decode("utf-8")
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/subprocess.py", line 395, in check_output
    **kwargs).stdout
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/subprocess.py", line 472, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/subprocess.py", line 775, in __init__
    restore_signals, start_new_session)
  File "/cvmfs/soft.computecanada.ca/easybuild/software/2017/Core/python/3.7.4/lib/python3.7/subprocess.py", line 1522, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
OSError: [Errno 14] Bad address: 'mpicc'
```
[3dgenerator.txt](https://github.com/firedrakeproject/firedrake/files/4712884/3dgenerator.txt)

**Comment by ScottMacLachlan** [comment]
The specific call that produced the above was fixed by branches in PRs https://github.com/firedrakeproject/firedrake/pull/1714 and https://github.com/OP2/PyOP2/pull/584

**Comment by wence-** [comment]
Was fixed by #1714.
